aip/auth/4115.md

```diff
@@ -5,124 +5,110 @@ state: approved
+been put in place. Therefore, using bound tokens is more secure than bearer tokens, which can be 
+stolen and adversarially replayed.
 
-Supporting the default credentials from Google virtual machines is considered as
-a part of the [Application Default Credential][2]. To understand the overall
-credential flow, please read [AIP-4110][2].
+This AIP describes the flow of:
 
-### Google Virtual Machine Environments
+1. Retrieving configurations through a metadata server (MDS) endpoint. The configuration contains 
+   required data to access Google’s default mTLS credentials.
+2. Requesting bound tokens.
 
-There are typically two types of Google virtual machine environments where the
-auth library should obtain the identity token based on the environment type:
+**Note:** Because this AIP describes guidance and requirements in a language-neutral way, it uses 
+generic terminology which may be imprecise or inappropriate in certain languages or environments.
 
-- Google App Engine Standard 1.0 (aka 1st generation)
-- Compute Engine 
-  - GCE equivalent runtimes
-    - Google App Engine Standard 2.0+
-    - Google App Engine Flex
-    - Google Cloud Functions
-    - Cloud Run
-    - Workload Identity on Google Kubernetes Engine
+## Prerequisites
 
-The auth library **should** depend on the [Google App Engine SDK][3] or well 
-defined GAE environment variables to detect if the application is running within
-the 1st generation Google App Engine environment.
+Before trying to use Google’s default mTLS credentials and bound tokens, the client **must** first 
+check if the remote Google API endpoint supports mTLS. If the remote endpoint does NOT support mTLS,
+the client **should** connect to the endpoint using TLS and request a token from MDS. How to check 
+if an endpoint supports mTLS is out of the scope of this AIP. If the remote endpoint does support 
+mTLS, the client **should** try to connect using mTLS first before falling back to TLS. How to find 
+the remote API’s mTLS endpoint is out of the scope of this AIP.
 
-To detect if the application is running on Compute Engine or an equivalent runtime,
-the auth library **should** depend on the [Metadata Service Library][4].
+## Guidance
 
-### Compute Engine or Equivalent Runtime
+### Access Default mTLS Credentials
 
-If the application runs on a Compute Engine instance (or equivalent runtime),
-the auth library can request the following URL to get an access token associated
-with instance:
+To leverage Google’s default mTLS credentials, the client **should** retrieve configurations from 
+MDS. All virtual environments (GCE, GAE and GKE) MDS expose a HTTP endpoint (mTLS configuration 
+endpoint) that serves mTLS configurations.
 
+The MDS mTLS configuration endpoint is: 
 ```
-http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token
+http://169.254.169.254/computeMetadata/v1/googleAutoMtlsConfiguration
```

Comment:
Generally speaking, I think we use the hostname here rather than the IP address?

Example: https://cloud.google.com/compute/docs/metadata/overview#limitations

Reply:
I remember someone telling me IP address is prefered to avoid some DNS attack, but since all these external docs still use hostnames, I changed it to hostname.